examples/seismic/source.py

```diff
@@ -1,60 +1,114 @@
-    :param time_order: (Optional) Time discretization order (defaults to 2)
+    :param coordinates: Point coordinates for this source.
+    :param time_range: (Optional) Tuple of time origin, time interval and
+                       number of timesteps, (t0, dt, ntime). Can only set
+                       one of time_range and time_values.
+    :param time_values: (Optional) Time values. Can only set one of time_range and time_values.
+    :param data: (Optional) Data values to initialise point data.
+    :param npoint: (Optional) Number of sparse points represented by this source.
+    :param time_order: (Optional) Time discretization order (defaults to 2).
     :param dimension: :(Optional) class:`Dimension` object for
-                       representing the number of points in this source
-
-    Note, either the dimensions `ntime` and `npoint` or the fully
-    initialised `data` array need to be provided.
+                       representing the number of points in this source.
     """
 
-    def __new__(cls, name, grid, ntime=None, npoint=None, data=None,
-                coordinates=None, **kwargs):
+    def __new__(cls, name, grid, time_range=None, time_values=None, npoint=None,
+                data=None, coordinates=None, **kwargs):
         p_dim = kwargs.get('dimension', Dimension(name='p_%s' % name))
         time_order = kwargs.get('time_order', 2)
         npoint = npoint or coordinates.shape[0]
-        if data is None:
-            if ntime is None:
-                error('Either data or ntime are required to'
-                      'initialise source/receiver objects')
+
+        if time_range is None and time_values is None:
+            error("Either time_range or time_values must be defined.")
+
+        Time =  namedtuple('Time', ['t0', 'dt', 'nt'])
+        if time_range is None:
+            cls._time_range = Time(time_values[0], time_values[1]-time_values[0], time_values.size)
         else:
-            ntime = ntime or data.shape[0]
+            cls._time_range = Time(*time_range)
+        nt = cls._time_range[2]
 
         # Create the underlying SparseTimeFunction object
         obj = SparseTimeFunction.__new__(cls, name=name, grid=grid,
                                          dimensions=[grid.time_dim, p_dim],
-                                         npoint=npoint, nt=ntime, time_order=time_order,
+                                         npoint=npoint, nt=nt,
+                                         time_order=time_order,
                                          coordinates=coordinates, **kwargs)
 
         # If provided, copy initial data into the allocated buffer
         if data is not None:
             obj.data[:] = data
+
         return obj
 
     def __init__(self, *args, **kwargs):
         if not self._cached():
             super(PointSource, self).__init__(*args, **kwargs)
 
+    @property
+    def _t_start(self):
+        return self._time_range[0]
```

Comment:
you can exploit the ``namedtuple`` now: you can use its fields rather than explicit indexing